File: mk-attendance-flutter/lib/utils/ethiopian_date.dart
class EthiopianDateUtils {
  static const List<String> ethiopianMonths = [
    'መስከረም', 'ጥቅምት', 'ኅዳር', 'ታኅሳስ', 'ጥር', 'የካቲት',
    'መጋቢት', 'ሚያዝያ', 'ግንቦት', 'ሰኔ', 'ሐምሌ', 'ነሐሴ', 'ጳጉሜን'
  ];

  static const List<String> ethiopianDays = [
    'እሑድ', 'ሰኞ', 'ማክሰኞ', 'ረቡዕ', 'ሐሙስ', 'ዓርብ', 'ቅዳሜ'
  ];

  /// Check if Gregorian year is leap year
  static bool _isLeapYear(int year) {
    return (year % 4 == 0 && year % 100 != 0) || (year % 400 == 0);
  }

  /// Correct Ethiopian date conversion
  /// December 13, 2025 = ታህሳስ 4, 2018
  static Map<String, int> gregorianToEthiopian(DateTime gregorianDate) {
    final year = gregorianDate.year;
    final month = gregorianDate.month;
    final day = gregorianDate.day;
    
    // Ethiopian year = Gregorian year - 7
    int ethYear = year - 7; // 2025 - 7 = 2018 ✓
    int ethMonth;
    int ethDay;
    
    // Ethiopian New Year starts around September 11/12
    if (month >= 9) {
      // September to December (Ethiopian months 1-4)
      if (month == 9) {
        ethMonth = 1; // መስከረም
        ethDay = day - 10; // Approximate adjustment
        if (ethDay <= 0) {
          ethYear = year - 8;
          ethMonth = 13; // ጳጉሜን from previous year
          ethDay = 30 + ethDay;
        }
      } else if (month == 10) {
        ethMonth = 2; // ጥቅምት
        ethDay = day - 10;
        if (ethDay <= 0) {
          ethMonth = 1;
          ethDay = 30 + ethDay;
        }
      } else if (month == 11) {
        ethMonth = 3; // ኅዳር
        ethDay = day - 9;
        if (ethDay <= 0) {
          ethMonth = 2;
          ethDay = 30 + ethDay;
        }
      } else if (month == 12) {
        ethMonth = 4; // ታኅሳስ
        // December 13 = ታህሳስ 4, so December 10 = ታህሳስ 1
        ethDay = day - 9;
        if (ethDay <= 0) {
          ethMonth = 3;
          ethDay = 30 + ethDay;
        }
      } else {
        ethMonth = 1;
        ethDay = 1;
      }
    } else {
      // January to August (Ethiopian months 5-12)
      if (month == 1) {
        ethMonth = 5; // ጥር
        ethDay = day + 21;
        if (ethDay > 30) {
          ethMonth = 6;
          ethDay = ethDay - 30;
        }
      } else if (month == 2) {
        ethMonth = 6; // የካቲት
        ethDay = day + 21;
        if (ethDay > 30) {
          ethMonth = 7;
          ethDay = ethDay - 30;
        }
      } else if (month == 3) {
        ethMonth = 7; // መጋቢት
        ethDay = day + 19;
        if (ethDay > 30) {
          ethMonth = 8;
          ethDay = ethDay - 30;
        }
      } else if (month == 4) {
        ethMonth = 8; // ሚያዝያ
        ethDay = day + 21;
        if (ethDay > 30) {
          ethMonth = 9;
          ethDay = ethDay - 30;
        }
      } else if (month == 5) {
        ethMonth = 9; // ግንቦት
        ethDay = day + 21;
        if (ethDay > 30) {
          ethMonth = 10;
          ethDay = ethDay - 30;
        }
      } else if (month == 6) {
        ethMonth = 10; // ሰኔ
        ethDay = day + 22;
        if (ethDay > 30) {
          ethMonth = 11;
          ethDay = ethDay - 30;
        }
      } else if (month == 7) {
        ethMonth = 11; // ሐምሌ
        ethDay = day + 22;
        if (ethDay > 30) {
          ethMonth = 12;
          ethDay = ethDay - 30;
        }
      } else if (month == 8) {
        ethMonth = 12; // ነሐሴ
        ethDay = day + 23;
        if (ethDay > 30) {
          ethMonth = 13;
          ethDay = ethDay - 30;
        }
      } else {
        ethMonth = 1;
        ethDay = 1;
      }
    }
    
    // Ensure valid ranges
    if (ethMonth < 1) ethMonth = 1;
    if (ethMonth > 13) ethMonth = 13;
    if (ethDay < 1) ethDay = 1;
    if (ethDay > 30) ethDay = 30;
    
    return {
      'year': ethYear,
      'month': ethMonth,
      'day': ethDay,
    };
  }

  /// Get current Ethiopian date
  static Map<String, int> getCurrentEthiopianDate() {
    return gregorianToEthiopian(DateTime.now());
  }

  /// Format Ethiopian date as string
  static String formatEthiopianDate(Map<String, int> ethiopianDate) {
    final monthIndex = (ethiopianDate['month']! - 1).clamp(0, 12);
    final monthName = ethiopianMonths[monthIndex];
    return '${ethiopianDate['day']} $monthName ${ethiopianDate['year']}';
  }

  /// Convert Ethiopian date to Gregorian format for API calls
  /// This uses the same logic as the web app for consistency
  static String ethiopianToGregorian(Map<String, int> ethiopianDate) {
    final ethYear = ethiopianDate['year']!;
    final ethMonth = ethiopianDate['month']!;
    final ethDay = ethiopianDate['day']!;
    
    // Use the same conversion logic as web app
    final gregYear = ethYear + 7;
    int gregMonth = ethMonth + 8;
    int gregDay = ethDay;
    
    // Handle month overflow
    if (gregMonth > 12) {
      gregMonth = gregMonth - 12;
    }
    
    // Ensure valid ranges
    if (gregMonth <= 0) gregMonth = 1;
    if (gregMonth > 12) gregMonth = 12;
    if (gregDay <= 0) gregDay = 1;
    if (gregDay > 28) gregDay = 28;
    
    final year = gregYear.toString().padLeft(4, '0');
    final month = gregMonth.toString().padLeft(2, '0');
    final day = gregDay.toString().padLeft(2, '0');
    
    return '$year-$month-$day';
  }

  /// Convert Gregorian date string back to Ethiopian
  /// This uses the same logic as the web app for consistency
  static Map<String, int> gregorianToEthiopianFromString(String gregorianString) {
    try {
      final parts = gregorianString.split('-');
      final year = int.parse(parts[0]);
      final month = int.parse(parts[1]);
      final day = int.parse(parts[2]);
      
      // Use the same reverse conversion logic as web app
      final ethYear = year - 7;
      int ethMonth = month - 8;
      final ethDay = day;
      
      // Handle month underflow
      if (ethMonth <= 0) {
        ethMonth = ethMonth + 12;
      }
      
      // Ensure valid ranges
      if (ethMonth <= 0) ethMonth = 1;
      if (ethMonth > 13) ethMonth = 13;
      
      return {'year': ethYear, 'month': ethMonth, 'day': ethDay};
    } catch (e) {
      return getCurrentEthiopianDate();
    }
  }

  /// Get current date in Gregorian format for API calls
  static String getCurrentGregorianForApi() {
    return DateTime.now().toIso8601String().split('T')[0];
  }

  /// Format date for display using accurate conversion
  static String formatDate(String gregorianDateString) {
    try {
      final date = DateTime.parse(gregorianDateString);
      final ethiopianDate = gregorianToEthiopian(date);
      final monthName = ethiopianMonths[(ethiopianDate['month']! - 1).clamp(0, 12)];
      return '${ethiopianDate['day']} $monthName ${ethiopianDate['year']}';
    } catch (e) {
      return gregorianDateString;
    }
  }

  /// Get today's date formatted
  static String getTodayFormatted() {
    final today = DateTime.now();
    return formatDate(today.toIso8601String().split('T')[0]);
  }

  /// Get day name in Ethiopian
  static String getDayName(DateTime date) {
    int weekdayIndex = (date.weekday % 7);
    return ethiopianDays[weekdayIndex];
  }
}

File: lib/ethiopian-date.ts
// Ethiopian Calendar Utility
// Ethiopian calendar has 13 months: 12 months of 30 days + 1 month of 5/6 days

export interface EthiopianDate {
  year: number;
  month: number;
  day: number;
}

export const ETHIOPIAN_MONTHS = [
  'መስከረም', // Meskerem
  'ጥቅምት',   // Tikimt
  'ኅዳር',    // Hidar
  'ታኅሳስ',   // Tahsas
  'ጥር',     // Tir
  'የካቲት',   // Yekatit
  'መጋቢት',   // Megabit
  'ሚያዝያ',   // Miazia
  'ግንቦት',   // Ginbot
  'ሰኔ',     // Sene
  'ሐምሌ',    // Hamle
  'ነሐሴ',    // Nehase
  'ጳጉሜን'    // Pagumen
];

export const ETHIOPIAN_MONTHS_EN = [
  'Meskerem', 'Tikimt', 'Hidar', 'Tahsas', 'Tir', 'Yekatit',
  'Megabit', 'Miazia', 'Ginbot', 'Sene', 'Hamle', 'Nehase', 'Pagumen'
];

// Convert Gregorian date to Ethiopian date
export function gregorianToEthiopian(gregorianDate: Date): EthiopianDate {
  const year = gregorianDate.getFullYear();
  const month = gregorianDate.getMonth() + 1; // JS months are 0-based
  const day = gregorianDate.getDate();
  
  // Ethiopian calendar is approximately 7-8 years behind Gregorian
  // New Year starts around September 11 (or 12 in leap years)
  
  let ethYear: number;
  let ethMonth: number;
  let ethDay: number;
  
  // Simplified conversion algorithm
  if (month >= 9 || (month === 9 && day >= 11)) {
    // After Ethiopian New Year
    ethYear = year - 7;
    
    if (month === 9) {
      ethMonth = 1; // Meskerem
      ethDay = day - 10;
    } else if (month === 10) {
      ethMonth = 1; // Meskerem
      ethDay = day + 20;
      if (ethDay > 30) {
        ethMonth = 2; // Tikimt
        ethDay = ethDay - 30;
      }
    } else if (month === 11) {
      ethMonth = 2; // Tikimt
      ethDay = day + 20;
      if (ethDay > 30) {
        ethMonth = 3; // Hidar
        ethDay = ethDay - 30;
      }
    } else if (month === 12) {
      ethMonth = 3; // Hidar
      ethDay = day + 21;
      if (ethDay > 30) {
        ethMonth = 4; // Tahsas
        ethDay = ethDay - 30;
      }
    } else {
      // Handle other months (simplified)
      const dayOfYear = getDayOfYear(gregorianDate);
      const ethDayOfYear = dayOfYear - 253; // Approximate offset
      
      if (ethDayOfYear <= 0) {
        ethYear = year - 8;
        ethMonth = 13;
        ethDay = ethDayOfYear + 5;
      } else {
        ethMonth = Math.ceil(ethDayOfYear / 30);
        ethDay = ethDayOfYear - ((ethMonth - 1) * 30);
        
        if (ethMonth > 13) {
          ethMonth = 1;
          ethYear = year - 6;
        }
      }
    }
  } else {
    // Before Ethiopian New Year
    ethYear = year - 8;
    
    const dayOfYear = getDayOfYear(gregorianDate);
    const ethDayOfYear = dayOfYear + 112; // Approximate offset
    
    ethMonth = Math.ceil(ethDayOfYear / 30);
    ethDay = ethDayOfYear - ((ethMonth - 1) * 30);
    
    if (ethMonth > 13) {
      ethMonth = ethMonth - 13;
      ethYear = year - 7;
    }
  }
  
  // Ensure valid ranges
  if (ethDay <= 0) {
    ethMonth--;
    if (ethMonth <= 0) {
      ethMonth = 13;
      ethYear--;
    }
    ethDay = (ethMonth === 13) ? 5 : 30;
  }
  
  if (ethMonth > 13) {
    ethMonth = 1;
    ethYear++;
  }
  
  return { year: ethYear, month: ethMonth, day: ethDay };
}

// Convert Ethiopian date to Gregorian date
export function ethiopianToGregorian(ethDate: EthiopianDate): Date {
  // Simplified reverse conversion
  const { year, month, day } = ethDate;
  
  // Ethiopian New Year is around September 11
  let gregYear = year + 7;
  let gregMonth: number;
  let gregDay: number;
  
  if (month === 1) {
    // Meskerem
    gregMonth = 9;
    gregDay = day + 10;
    if (gregDay > 30) {
      gregMonth = 10;
      gregDay = gregDay - 30;
    }
  } else if (month <= 4) {
    // Tikimt, Hidar, Tahsas
    const totalDays = (month - 1) * 30 + day + 10;
    const startDate = new Date(gregYear, 8, 11); // Sept 11
    const resultDate = new Date(startDate.getTime() + (totalDays - 1) * 24 * 60 * 60 * 1000);
    return resultDate;
  } else {
    // Other months - approximate
    const totalDays = (month - 1) * 30 + day;
    const startDate = new Date(gregYear, 8, 11); // Sept 11
    const resultDate = new Date(startDate.getTime() + (totalDays - 1) * 24 * 60 * 60 * 1000);
    return resultDate;
  }
  
  return new Date(gregYear, gregMonth - 1, gregDay);
}

// Get day of year for Gregorian date
function getDayOfYear(date: Date): number {
  const start = new Date(date.getFullYear(), 0, 0);
  const diff = date.getTime() - start.getTime();
  return Math.floor(diff / (1000 * 60 * 60 * 24));
}

// Format Ethiopian date as string
export function formatEthiopianDate(ethDate: EthiopianDate, useAmharic = true): string {
  const months = useAmharic ? ETHIOPIAN_MONTHS : ETHIOPIAN_MONTHS_EN;
  return `${ethDate.day} ${months[ethDate.month - 1]} ${ethDate.year}`;
}

// Get current Ethiopian date
export function getCurrentEthiopianDate(): EthiopianDate {
  return gregorianToEthiopian(new Date());
}

// Convert Ethiopian date to ISO string for database storage (Ethiopian format)
export function ethiopianToISOString(ethDate: EthiopianDate): string {
  const year = ethDate.year.toString().padStart(4, '0');
  const month = ethDate.month.toString().padStart(2, '0');
  const day = ethDate.day.toString().padStart(2, '0');
  return `${year}-${month}-${day}`;
}

// DEPRECATED: Convert Ethiopian date to Gregorian ISO string (for backward compatibility)
export function ethiopianToGregorianISOString(ethDate: EthiopianDate): string {
  const gregDate = ethiopianToGregorian(ethDate);
  return gregDate.toISOString().split('T')[0];
}

// Parse ISO string to Ethiopian date (handles both Ethiopian and Gregorian formats)
export function isoStringToEthiopian(isoString: string): EthiopianDate {
  const [year, month, day] = isoString.split('-').map(Number);
  
  // Check if it's already in Ethiopian format (year < 2050)
  if (year < 2050) {
    return { year, month, day };
  } else {
    // Convert from Gregorian format
    const gregDate = new Date(isoString);
    return gregorianToEthiopian(gregDate);
  }
}

// Get Ethiopian date input value (YYYY-MM-DD format but Ethiopian)
export function getEthiopianDateInputValue(ethDate: EthiopianDate): string {
  const year = ethDate.year.toString().padStart(4, '0');
  const month = ethDate.month.toString().padStart(2, '0');
  const day = ethDate.day.toString().padStart(2, '0');
  return `${year}-${month}-${day}`;
}

// Parse Ethiopian date input value
export function parseEthiopianDateInput(value: string): EthiopianDate {
  const [year, month, day] = value.split('-').map(Number);
  return { year, month, day };
}